### PR TITLE
feat: Pydantic support

### PR DESCRIPTION
**Pull request by dannygoldstein**
Adds pydantic support to weave. 

* typeof(any pydantic model) -> RelocatableObjectType(attrs)
* save(any pydantic model) -> .get() node 
* weave.use(weave.save(any pydantic model)) -> an object with the right attributes

### CODE REVIEW

File: weave/weave_types.py [hunk 1]
```diff
@@ -1075,6 +1092,12 @@ def load_instance(self, artifact, name, extra=None):
         mapper = mappers_python.map_from_python(self, artifact)
         return mapper.apply(result)
 
+    def __eq__(self, other) -> bool:
```
Comment (shawnlewis) [inline weave/weave_types.py:1095]: Hmm, why did you need this? I thought we had implemented __hash__ everywhere which made all types comparable with __eq__ but maybe I'm wrong about that.

File: weave/ops_domain/wbmedia.py [hunk 2]
```diff
@@ -11,10 +11,18 @@
 from ..ops_primitives import markdown
 
 
+class LegacyImageArtifactFileRef:
+    # Dummy class to make LegacyImageArtifactFileRefType not have
+    # pydantic.BaseModel as an instance class, which messes up type_of.
```
Comment (shawnlewis) [inline weave/ops_domain/wbmedia.py:16]: how does it mess up type_of?

**Review by shawnlewis — COMMENTED** [review]
Super cool

File: weave/weave_types.py (hunk 1 above)
Comment (dannygoldstein) [reply weave/weave_types.py:1095]: It's because all `ObjectType`s are currently equal in weave,

```
In [2]: from weave import weave_types as types

In [3]: ot1 = types.ObjectType(a=types.Int())

In [4]: ot2 = types.ObjectType(b=types.String())

In [5]: ot1 == ot2
Out[5]: True
```

on master.

File: weave/ops_domain/wbmedia.py (hunk 2 above)
Comment (dannygoldstein) [reply weave/ops_domain/wbmedia.py:16]: Without this, we get 

```
____________________________________________________________ test_pydantic_type_inference ____________________________________________________________

    def test_pydantic_type_inference():
        class TestPydanticClass(pydantic.BaseModel):
            a: int
            b: str
            c: float
    
        obj = TestPydanticClass(a=1, b="2", c=3.0)
>       t = types.TypeRegistry.type_of(obj)

test_pydantic.py:52: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../weave_types.py:158: in type_of
    obj_type = type_.type_of(obj)
../weave_types.py:310: in type_of
    return cls.type_of_instance(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'weave.ops_domain.wbmedia.LegacyImageArtifactFileRefType'>, obj = TestPydanticClass(a=1, b='2', c=3.0)

    @classmethod
    def type_of_instance(cls, obj):
>       raise errors.WeaveTypeError(
            "LegacyImageArtifactFileRefType should never be used! it is a temp type"
        )
E       weave.errors.WeaveTypeError: LegacyImageArtifactFileRefType should never be used! it is a temp type

../ops_domain/wbmedia.py:25: WeaveTypeError
```
